src/main.rs

```diff
@@ -342,3 +307,75 @@ fn err_to_warp(e: podsync::Error) -> impl warp::Reply {
 pub struct QuerySince {
     since: crate::time::Timestamp,
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum UsernameFormat {
+    Name,
+    NameJson,
+}
+impl UsernameFormat {
+    pub fn convert<'a>(&self, username: &'a str) -> Result<&'a str, warp::Rejection> {
+        match self {
+            Self::Name => Ok(username),
+            Self::NameJson => split_format_json(username).map_err(warp::reject::custom),
+        }
+    }
+}
+
+fn cookie_authorize(
+    username_fmt: UsernameFormat,
+    podsync: Arc<PodSync>,
+) -> impl Filter<Extract = (PodSyncAuthed<true>,), Error = warp::Rejection> + Clone {
+    let cookie_auth_check = warp::cookie(COOKIE_NAME).and_then({
+        move |session_id: SessionId| {
+            let podsync = Arc::clone(&podsync);
+
+            async move {
+                podsync
+                    .authenticate(session_id)
+                    .await
+                    .map_err(warp::reject::custom)
+            }
+        }
+    });
+
+    warp::path::param::<String>()
+        .and(cookie_auth_check)
+        .and_then(move |username: String, podsync: PodSyncAuthed| async move {
+            podsync
+                .with_user(username_fmt.convert(&username)?)
+                .map_err(warp::reject::custom)
+        })
+}
+
+fn login_authorize(
+    username_fmt: UsernameFormat,
+    podsync: Arc<PodSync>,
+) -> impl Filter<Extract = (PodSyncAuthed<true>,), Error = warp::Rejection> + Clone {
+    warp::path::param::<String>()
+        .and(warp::header("authorization"))
+        .and(warp::cookie::optional(COOKIE_NAME))
+        .and_then(
+            move |username: String, auth: BasicAuth, session_id: Option<SessionId>| {
+                let podsync = Arc::clone(&podsync);
+                async move {
+                    let username = username_fmt.convert(&username)?;
+                    let auth = auth.with_path_username(&username)?;
+                    podsync
+                        .login(auth, session_id)
+                        .await?
+                        .with_user(&username)
+                        .map_err(warp::reject::custom)
```

Comment:
This is the same situation as [above](https://github.com/bobrippling/podsync/pull/2/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR347) for the cookie auth - we'll want to pass back a 4xx since the client's attempted to login with the wrong password (again, feel free to have a go if you like)